src/thread_event.c

```diff
@@ -0,0 +1,252 @@
+	uint64_t wait = next_event - last_event;
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	assert(thread_allocated - last_event < wait);
+
+	bool no_event_on;
+	uint64_t min_wait = thread_allocated_next_event_compute(tsd,
+	    &no_event_on);
+
+	if (no_event_on) {
+		/*
+		 * next_event should have been pushed up only except when the
+		 * TSD is just initialized.  The last_event == 0 guard below is
+		 * stronger than needed, but having an exactly accurate guard
+		 * is more complicated to implement.
+		 */
+		assert(last_event == 0 ||
+		    next_event == THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX ||
+		    next_event == THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX + 1);
+	} else {
+		assert(min_wait == wait);
+	}
+}
+
+static void
+thread_event_adjust_thresholds_helper(tsd_t *tsd, uint64_t next_event) {
+	thread_allocated_next_event_set(tsd, next_event);
+	uint64_t next_event_fast = next_event <=
+	    THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX ? next_event : 0U;
+	thread_allocated_next_event_fast_set(tsd, next_event_fast);
+}
+
+void
+thread_event_trigger(tsd_t *tsd, uint64_t usize, bool delay_event) {
+	/* usize has already been added to thread_allocated. */
+	uint64_t thread_allocated_after = thread_allocated_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t accumbytes = thread_allocated_after -
+	    thread_allocated_last_event_get(tsd);
+
+	if (accumbytes < usize) {
+		/* accumbytes overflowed. */
+		accumbytes = UINT64_MAX;
+	}
+
+	uint64_t wait = UINT64_MAX;
+	bool no_event_on = true;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		no_event_on = false;					\
+		uint64_t event_wait = event##_event_wait_get(tsd);	\
+		assert(event_wait > 0);					\
+		if (event_wait <= accumbytes) {				\
+			event_wait = 0;					\
+		} else {						\
+			event_wait -= accumbytes;			\
+		}							\
+		if ((delay_event || !tsd_nominal(tsd) ||		\
+		    tsd_reentrancy_level_get(tsd) > 0) &&		\
+		    event_wait == 0) {					\
+			/* Delay the event. */				\
+			event_wait = 1;					\
+		}							\
+		event##_event_wait_set(tsd, event_wait);		\
+		/*							\
+		 * If there is a single event, then the remaining wait	\
+		 * time won't be positive, and we rely on a		\
+		 * thread_event_update() call later to properly set the	\
+		 * next next_event; if there are multiple events, then	\
+		 * here we can get the minimum remaining wait time to	\
+		 * the next already set event.				\
+		 */							\
+		if (event_wait > 0 && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	/*
+	 * In case none of the events has its condition on, set next_event to
+	 * be THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX to minimize malloc() slow
+	 * path triggering.
+	 */
+	if (no_event_on) {
+		/* The subtraction is intentionally susceptible to underflow. */
+		wait = THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX -
+		    thread_allocated_after;
+		if (wait == 0U) {
```

Comment:
nit: update to

    if (wait < MIN_WAIT) {
        wait = MIN_WAIT;
    }

I'd say also remove the comments below. Might cause more confusion than it helps here.